flink-clients/src/test/java/org/apache/flink/client/program/PackagedProgramRetrieverImplTest.java

```diff
@@ -0,0 +1,573 @@
+    }
+
+    // TODO: this test checks the same code path as
+    // testDeriveEntryClassInformationFromSystemClasspathWithNonExistingJobClassName
+    // We should make it fail early if the class is not present on the system classpath
+    // Right now, the test is failing because no error is thrown
+    @Ignore
+    @Test(expected = FlinkException.class)
+    public void testEntryClassNotFoundOnSystemClasspath() throws FlinkException {
+        PackagedProgramRetrieverImpl.create(
+                null, "NotExistingClass", new String[0], new Configuration());
+    }
+
+    @Test(expected = FlinkException.class)
+    public void testEntryClassNotFoundOnUserClasspath() throws FlinkException {
+        PackagedProgramRetrieverImpl.create(
+                classpathProvider.getUserDirectoryWithNoEntryClass(),
+                classpathProvider.getTestJobClassName(),
+                new String[0],
+                new Configuration());
+    }
+
+    @Test(expected = FlinkException.class)
+    public void testTooManyEntryClassesOnSystemClasspath() throws FlinkException {
+        PackagedProgramRetrieverImpl.create(null, null, new String[0], new Configuration());
+    }
+
+    @Test
+    public void testTooManyEntryClassesOnUserClasspath() throws FlinkException, IOException {
+        classpathProvider.setSystemClasspathWithTwoEntryClasses();
+        final PackagedProgramRetrieverImpl retriever =
+                PackagedProgramRetrieverImpl.create(
+                        classpathProvider.getUserDirectoryWithTwoEntryClasses(),
+                        classpathProvider.getTestJobClassName(),
+                        new String[0],
+                        new Configuration());
+        assertThat(
+                retriever.getPackagedProgram().getMainClassName(),
+                is(classpathProvider.getTestJobClassName()));
+    }
+
+    @Test
+    public void testRetrieveCorrectUserClasspathsWithoutSpecifiedEntryClass()
+            throws IOException, FlinkException, ProgramInvocationException {
+        final PackagedProgramRetriever retrieverUnderTest =
+                PackagedProgramRetrieverImpl.create(
+                        classpathProvider.getUserDirectoryWithOneEntryClass(),
+                        null,
+                        classpathProvider.getTestJobArgs("suffix"),
+                        new Configuration());
+        final JobGraph jobGraph = retrieveJobGraph(retrieverUnderTest, new Configuration());
+        final List<String> actualClasspath =
+                jobGraph.getClasspaths().stream().map(URL::toString).collect(Collectors.toList());
+
+        final Path workingDirectory = FileUtils.getCurrentWorkingDirectory();
+        final List<String> expectedClasspath = new ArrayList<>();
+        for (File file : classpathProvider.getUserDirectoryWithOneEntryClass().listFiles()) {
+            if (!file.getName().endsWith("jar")) {
+                // only jars are expected
+                continue;
+            }
```

Comment:
Why should we silently ignore this? Wouldn't it be a bug in the ClassPathProvider?

Reply:
The `ClasspathProvider` provides all the resources that are on the classpath (including a text file in most of the cases). The production code filters them out in [PackagedProgramRetrieverImpl.extractUserClasspath](https://github.com/XComp/flink/blob/e9a2e57adf4f45958e6fa0e9e9f0f7afac50d01e/flink-clients/src/main/java/org/apache/flink/client/program/PackagedProgramRetrieverImpl.java#L208). The filtering here is just the corresponding test code checking this filtering.

Reply:
But I updated the condition to use `FileUtils` and added a more meaningful comment